wrapper/src/main/java/software/amazon/jdbc/plugin/customendpoint/CustomEndpointPlugin.java

```diff
@@ -0,0 +1,370 @@
+   * @param props         The properties that the custom endpoint plugin should use.
+   * @param rdsClientFunc The function to call to obtain an {@link RdsClient} instance.
+   */
+  public CustomEndpointPlugin(
+      final PluginService pluginService,
+      final Properties props,
+      final BiFunction<HostSpec, Region, RdsClient> rdsClientFunc) {
+    this.pluginService = pluginService;
+    this.props = props;
+    this.rdsClientFunc = rdsClientFunc;
+
+    this.waitOnCachedInfoDurationMs = WAIT_FOR_CUSTOM_ENDPOINT_INFO_TIMEOUT_MS.getInteger(this.props);
+    this.idleMonitorExpirationMs = CUSTOM_ENDPOINT_MONITOR_IDLE_EXPIRATION_MS.getInteger(this.props);
+
+    TelemetryFactory telemetryFactory = pluginService.getTelemetryFactory();
+    this.endpointConnectionsCounter = telemetryFactory.createCounter(TELEMETRY_CUSTOM_ENDPOINT_COUNTER);
+    this.waitForInfoCounter = telemetryFactory.createCounter(TELEMETRY_WAIT_FOR_INFO_COUNTER);
+  }
+
+  @Override
+  public Set<String> getSubscribedMethods() {
+    return subscribedMethods;
+  }
+
+  @Override
+  public Connection connect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> connectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, connectFunc);
+  }
+
+  @Override
+  public Connection forceConnect(
+      final String driverProtocol,
+      final HostSpec hostSpec,
+      final Properties props,
+      final boolean isInitialConnection,
+      final JdbcCallable<Connection, SQLException> forceConnectFunc)
+      throws SQLException {
+    return connectInternal(hostSpec, props, forceConnectFunc);
+  }
+
+  /**
+   * Establishes a connection based on the passed in parameters. If the connection being made is to a custom endpoint
+   * URL, a monitor for that custom endpoint will be created if it does not already exist.
+   *
+   * @param hostSpec    The host information for the desired connection.
+   * @param props       The connection properties.
+   * @param connectFunc The connect pipeline function to call to establish the connection.
+   * @return The connection established by the {@code connectFunc}.
+   * @throws SQLException If an error occurs while attempting to establish a connection, or if we are connecting to a
+   *                      custom endpoint and we time out while waiting for custom endpoint info to be placed in the
+   *                      cache by the custom endpoint monitor.
+   */
+  protected Connection connectInternal(
+      HostSpec hostSpec,
+      Properties props,
+      JdbcCallable<Connection, SQLException> connectFunc) throws SQLException {
+    if (!this.rdsUtils.isRdsCustomClusterDns(hostSpec.getHost())) {
+      return connectFunc.call();
+    }
+
+    this.customEndpointHostSpec = hostSpec;
+    this.endpointConnectionsCounter.inc();
```

Comment:
I'm not sure what the value this counter can provide to a user.

Reply:
It would allow them to see how many connections are being made to custom endpoints specifically. If you think its not valuable I'll remove it.